The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/24232064165/job/70745780406) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/70745780406) [(plain)](https://github.com/rust-lang/rust/commit/0753fe08d6d1f5585dbb930ebb667b331a947b7e/checks/70745780406/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
fmt: checked 6786 files
tidy check
tidy [rustdoc_json (src)]: `rustdoc-json-types` modified, checking format version
tidy: Skipping binary file check, read-only filesystem
tidy [style (library)]: /checkout/library/core/src/mem/mod.rs:1144: undocumented unsafe
tidy [style (library)]: /checkout/library/core/src/mem/mod.rs:1148: undocumented unsafe
tidy [style (library)]: /checkout/library/core/src/mem/mod.rs:1149: undocumented unsafe
tidy [style (library)]: /checkout/library/core/src/mem/mod.rs:1180: undocumented unsafe
tidy [style (library)]: FAIL
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (26.0.1)
linting python files
---
linting javascript files
Running eslint on rustdoc JS files
info: ES-Check: there were no ES version matching errors!  🎉
typechecking javascript files
tidy: The following check failed: style (library)
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy --root-path=/checkout --cargo-path=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo --output-dir=/checkout/obj/build --concurrency=4 --npm-path=/node/bin/yarn --ci=true --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1618:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1417:29

--- BACKTRACE vvv
   0: <bootstrap::utils::exec::DeferredCommand>::finish_process
             at /checkout/src/bootstrap/src/utils/exec.rs:939:17
   1: <bootstrap::utils::exec::DeferredCommand>::wait_for_output::<&bootstrap::utils::exec::ExecutionContext>
             at /checkout/src/bootstrap/src/utils/exec.rs:831:21
   2: <bootstrap::utils::exec::ExecutionContext>::run
             at /checkout/src/bootstrap/src/utils/exec.rs:741:45
   3: <bootstrap::utils::exec::BootstrapCommand>::run::<&bootstrap::core::builder::Builder>
             at /checkout/src/bootstrap/src/utils/exec.rs:339:27
   4: <bootstrap::core::build_steps::test::Tidy as bootstrap::core::builder::Step>::run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:1417:29
   5: <bootstrap::core::builder::Builder>::ensure::<bootstrap::core::build_steps::test::Tidy>
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1579:36
   6: <bootstrap::core::build_steps::test::Tidy as bootstrap::core::builder::Step>::make_run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:1339:21
   7: <bootstrap::core::builder::StepDescription>::maybe_run
             at /checkout/src/bootstrap/src/core/builder/mod.rs:476:13
   8: bootstrap::core::builder::cli_paths::match_paths_to_steps_and_run
             at /checkout/src/bootstrap/src/core/builder/cli_paths.rs:232:18
   9: <bootstrap::core::builder::Builder>::run_step_descriptions
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1122:9
  10: <bootstrap::core::builder::Builder>::execute_cli
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1101:14
  11: <bootstrap::Build>::build
             at /checkout/src/bootstrap/src/lib.rs:799:25
  12: bootstrap::main
             at /checkout/src/bootstrap/src/bin/main.rs:130:11
  13: <fn() as core::ops::function::FnOnce<()>>::call_once
             at /rustc/ad726b5063362ec9897ef3d67452fc5606ee70fa/library/core/src/ops/function.rs:250:5
  14: std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>
             at /rustc/ad726b5063362ec9897ef3d67452fc5606ee70fa/library/std/src/sys/backtrace.rs:166:18
  15: std::rt::lang_start::<()>::{closure#0}
             at /rustc/ad726b5063362ec9897ef3d67452fc5606ee70fa/library/std/src/rt.rs:206:18
  16: <&dyn core::ops::function::Fn<(), Output = i32> + core::panic::unwind_safe::RefUnwindSafe + core::marker::Sync as core::ops::function::FnOnce<()>>::call_once
             at /rustc/ad726b5063362ec9897ef3d67452fc5606ee70fa/library/core/src/ops/function.rs:287:21
  17: std::panicking::catch_unwind::do_call::<&dyn core::ops::function::Fn<(), Output = i32> + core::panic::unwind_safe::RefUnwindSafe + core::marker::Sync, i32>
             at /rustc/ad726b5063362ec9897ef3d67452fc5606ee70fa/library/std/src/panicking.rs:581:40
  18: std::panicking::catch_unwind::<i32, &dyn core::ops::function::Fn<(), Output = i32> + core::panic::unwind_safe::RefUnwindSafe + core::marker::Sync>
             at /rustc/ad726b5063362ec9897ef3d67452fc5606ee70fa/library/std/src/panicking.rs:544:19
  19: std::panic::catch_unwind::<&dyn core::ops::function::Fn<(), Output = i32> + core::panic::unwind_safe::RefUnwindSafe + core::marker::Sync, i32>
             at /rustc/ad726b5063362ec9897ef3d67452fc5606ee70fa/library/std/src/panic.rs:359:14
  20: std::rt::lang_start_internal::{closure#0}
             at /rustc/ad726b5063362ec9897ef3d67452fc5606ee70fa/library/std/src/rt.rs:175:24
  21: std::panicking::catch_unwind::do_call::<std::rt::lang_start_internal::{closure#0}, isize>
             at /rustc/ad726b5063362ec9897ef3d67452fc5606ee70fa/library/std/src/panicking.rs:581:40
---
  28: __libc_start_main
  29: _start


Command has failed. Rerun with -v to see more details.
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Build completed unsuccessfully in 0:02:56
  local time: Fri Apr 10 07:41:37 UTC 2026
  network time: Fri, 10 Apr 2026 07:41:37 GMT
##[error]Process completed with exit code 1.
```

</details>

--- a/library/core/src/mem/mod.rs
+++ b/library/core/src/mem/mod.rs
@@ -32,6 +32,7 @@
 
 use crate::alloc::Layout;
 use crate::clone::TrivialClone;
+use crate::cmp::Ordering;
 use crate::marker::{Destruct, DiscriminantKind};
 use crate::panic::const_assert;
 use crate::{clone, cmp, fmt, hash, intrinsics, ptr};
@@ -1086,6 +1087,102 @@ pub const unsafe fn transmute_copy<Src, Dst>(src: &Src) -> Dst {
         // The caller must guarantee that the actual transmutation is safe.
         unsafe { ptr::read(src as *const Src as *const Dst) }
     }
+}
+
+/// Like [`transmute`], but only initializes the "common prefix" of the first
+/// `min(size_of::<Src>(), size_of::<Dst>())` bytes of the destination from the
+/// corresponding bytes of the source.
+///
+/// This is equivalent to a "union cast" through a `union` with `#[repr(C)]`.
+///
+/// That means some size mismatches are not UB, like `[T; 2]` to `[T; 1]`.
+/// Increasing size is usually UB from being insufficiently initialized -- like
+/// `u8` to `u32` -- but isn't always.  For example, going from `u8` to
+/// `#[repr(C, align(4))] AlignedU8(u8);` is sound.
+///
+/// Prefer normal `transmute` where possible, for the extra checking, since
+/// both do exactly the same thing at runtime, if they both compile.
+///
+/// # Safety
+///
+/// If `size_of::<Src>() >= size_of::<Dst>()`, the first `size_of::<Dst>()` bytes
+/// of `src` must be be *valid* when interpreted as a `Dst`.  (In this case, the
+/// preconditions are the same as for `transmute_copy(&ManuallyDrop::new(src))`.)
+///
+/// If `size_of::<Src>() <= size_of::<Dst>()`, the bytes of `src` padded with
+/// uninitialized bytes afterwards up to a total size of `size_of::<Dst>()`
+/// must be *valid* when interpreted as a `Dst`.
+///
+/// In both cases, any safety preconditions of the `Dst` type must also be upheld.
+///
+/// # Examples
+///
+/// ```
+/// #![feature(transmute_prefix)]
+/// use std::mem::transmute_prefix;
+///
+/// assert_eq!(unsafe { transmute_prefix::<[i32; 4], [i32; 2]>([1, 2, 3, 4]) }, [1, 2]);
+///
+/// let expected = if cfg!(target_endian = "little") { 0x34 } else { 0x12 };
+/// assert_eq!(unsafe { transmute_prefix::<u16, u8>(0x1234) }, expected);
+///
+/// // Would be UB because the destination is incompletely initialized.
+/// // transmute_prefix::<u8, u16>(123)
+///
+/// // OK because the destination is allowed to be partially initialized.
+/// let _: std::mem::MaybeUninit<u16> = unsafe { transmute_prefix(123_u8) };
+/// ```
+#[unstable(feature = "transmute_prefix", issue = "155079")]
+pub const unsafe fn transmute_prefix<Src, Dst>(src: Src) -> Dst {
+    #[repr(C)]
+    union Transmute<A, B> {
+        a: ManuallyDrop<A>,
+        b: ManuallyDrop<B>,
+    }
+
+    match const { Ord::cmp(&Src::SIZE, &Dst::SIZE) } {
+        // SAFETY: When Dst is bigger, the union is the size of Dst
+        Ordering::Less => unsafe {
+            let a = transmute_neo(src);
+            intrinsics::transmute_unchecked(Transmute::<Src, Dst> { a })
+        },
+        // SAFETY: When they're the same size, we can use the MIR primitive
+        Ordering::Equal => unsafe { intrinsics::transmute_unchecked::<Src, Dst>(src) },
+        // SAFETY: When Src is bigger, the union is the size of Src
+        Ordering::Greater => unsafe {
+            let u: Transmute<Src, Dst> = intrinsics::transmute_unchecked(src);
+            transmute_neo(u.b)
+        },
+    }
+}
+
+/// New version of `transmute`, exposed under this name so it can be iterated upon
+/// without risking breakage to uses of "real" transmute.
+///
+/// It will not be stabilized under this name.
+///
+/// # Examples
+///
+/// ```
+/// #![feature(transmute_neo)]
+/// use std::mem::transmute_neo;
+///
+/// assert_eq!(unsafe { transmute_neo::<f32, u32>(0.0) }, 0);
+/// ```
+///
+/// ```compile_fail,E0080
+/// #![feature(transmute_neo)]
+/// use std::mem::transmute_neo;
+///
+/// unsafe { transmute_neo::<u32, u16>(123) };
+/// ```
+#[unstable(feature = "transmute_neo", issue = "155079")]
+pub const unsafe fn transmute_neo<Src, Dst>(src: Src) -> Dst {
+    const { assert!(Src::SIZE == Dst::SIZE) };
+
+    // SAFETY: the const-assert just checked that they're the same size,
+    // and any other safety invariants need to be upheld by the caller.
+    unsafe { intrinsics::transmute_unchecked(src) }
 }
 
 /// Opaque type representing the discriminant of an enum.

--- a/tests/mir-opt/pre-codegen/transmutes.forget_at_home.PreCodegen.after.mir
+++ b/tests/mir-opt/pre-codegen/transmutes.forget_at_home.PreCodegen.after.mir
@@ -1,0 +1,20 @@
+// MIR for `forget_at_home` after PreCodegen
+
+fn forget_at_home(_1: String) -> () {
+    debug x => _1;
+    let mut _0: ();
+    scope 1 (inlined transmute_prefix::<String, ()>) {
+        scope 2 {
+        }
+        scope 3 {
+            scope 4 (inlined transmute_neo::<ManuallyDrop<()>, ()>) {
+            }
+        }
+        scope 5 (inlined transmute_neo::<String, ManuallyDrop<String>>) {
+        }
+    }
+
+    bb0: {
+        return;
+    }
+}

--- a/tests/mir-opt/pre-codegen/transmutes.neo_to_cast.PreCodegen.after.mir
+++ b/tests/mir-opt/pre-codegen/transmutes.neo_to_cast.PreCodegen.after.mir
@@ -1,0 +1,13 @@
+// MIR for `neo_to_cast` after PreCodegen
+
+fn neo_to_cast(_1: f32) -> i32 {
+    debug x => _1;
+    let mut _0: i32;
+    scope 1 (inlined transmute_neo::<f32, i32>) {
+    }
+
+    bb0: {
+        _0 = copy _1 as i32 (Transmute);
+        return;
+    }
+}

--- a/tests/mir-opt/pre-codegen/transmutes.pad_for_alignment.PreCodegen.after.mir
+++ b/tests/mir-opt/pre-codegen/transmutes.pad_for_alignment.PreCodegen.after.mir
@@ -1,0 +1,29 @@
+// MIR for `pad_for_alignment` after PreCodegen
+
+fn pad_for_alignment(_1: u32) -> Align64<u32> {
+    debug x => _1;
+    let mut _0: Align64<u32>;
+    scope 1 (inlined transmute_prefix::<u32, Align64<u32>>) {
+        let _2: std::mem::ManuallyDrop<u32>;
+        let mut _3: std::mem::transmute_prefix::Transmute<u32, Align64<u32>>;
+        scope 2 {
+        }
+        scope 3 {
+            scope 4 (inlined transmute_neo::<ManuallyDrop<Align64<u32>>, Align64<u32>>) {
+            }
+        }
+        scope 5 (inlined transmute_neo::<u32, ManuallyDrop<u32>>) {
+        }
+    }
+
+    bb0: {
+        StorageLive(_2);
+        _2 = copy _1 as std::mem::ManuallyDrop<u32> (Transmute);
+        StorageLive(_3);
+        _3 = transmute_prefix::Transmute::<u32, Align64<u32>> { a: copy _2 };
+        _0 = move _3 as Align64<u32> (Transmute);
+        StorageDead(_3);
+        StorageDead(_2);
+        return;
+    }
+}

--- a/tests/mir-opt/pre-codegen/transmutes.prefix_of_array.PreCodegen.after.mir
+++ b/tests/mir-opt/pre-codegen/transmutes.prefix_of_array.PreCodegen.after.mir
@@ -1,0 +1,29 @@
+// MIR for `prefix_of_array` after PreCodegen
+
+fn prefix_of_array(_1: [u32; 4]) -> [u32; 2] {
+    debug x => _1;
+    let mut _0: [u32; 2];
+    scope 1 (inlined transmute_prefix::<[u32; 4], [u32; 2]>) {
+        let _2: std::mem::transmute_prefix::Transmute<[u32; 4], [u32; 2]>;
+        let mut _3: std::mem::ManuallyDrop<[u32; 2]>;
+        scope 2 {
+        }
+        scope 3 {
+            scope 4 (inlined transmute_neo::<ManuallyDrop<[u32; 2]>, [u32; 2]>) {
+            }
+        }
+        scope 5 (inlined transmute_neo::<[u32; 4], ManuallyDrop<[u32; 4]>>) {
+        }
+    }
+
+    bb0: {
+        StorageLive(_2);
+        _2 = copy _1 as std::mem::transmute_prefix::Transmute<[u32; 4], [u32; 2]> (Transmute);
+        StorageLive(_3);
+        _3 = move (_2.1: std::mem::ManuallyDrop<[u32; 2]>);
+        _0 = copy _3 as [u32; 2] (Transmute);
+        StorageDead(_3);
+        StorageDead(_2);
+        return;
+    }
+}

--- a/tests/mir-opt/pre-codegen/transmutes.prefix_to_cast.PreCodegen.after.mir
+++ b/tests/mir-opt/pre-codegen/transmutes.prefix_to_cast.PreCodegen.after.mir
@@ -1,0 +1,21 @@
+// MIR for `prefix_to_cast` after PreCodegen
+
+fn prefix_to_cast(_1: f32) -> i32 {
+    debug x => _1;
+    let mut _0: i32;
+    scope 1 (inlined transmute_prefix::<f32, i32>) {
+        scope 2 {
+        }
+        scope 3 {
+            scope 4 (inlined transmute_neo::<ManuallyDrop<i32>, i32>) {
+            }
+        }
+        scope 5 (inlined transmute_neo::<f32, ManuallyDrop<f32>>) {
+        }
+    }
+
+    bb0: {
+        _0 = copy _1 as i32 (Transmute);
+        return;
+    }
+}

--- a/tests/mir-opt/pre-codegen/transmutes.rs
+++ b/tests/mir-opt/pre-codegen/transmutes.rs
@@ -1,0 +1,50 @@
+//@ compile-flags: -O -Zmir-opt-level=2
+
+#![crate_type = "lib"]
+#![feature(transmute_neo)]
+#![feature(transmute_prefix)]
+
+use std::mem::{transmute_neo, transmute_prefix};
+
+// EMIT_MIR transmutes.neo_to_cast.PreCodegen.after.mir
+pub fn neo_to_cast(x: f32) -> i32 {
+    // CHECK-LABEL: fn neo_to_cast
+    // CHECK: _0 = copy _1 as i32 (Transmute);
+    unsafe { transmute_neo(x) }
+}
+
+// EMIT_MIR transmutes.prefix_to_cast.PreCodegen.after.mir
+pub fn prefix_to_cast(x: f32) -> i32 {
+    // CHECK-LABEL: fn prefix_to_cast
+    // CHECK: _0 = copy _1 as i32 (Transmute);
+    unsafe { transmute_prefix(x) }
+}
+
+// EMIT_MIR transmutes.prefix_of_array.PreCodegen.after.mir
+pub fn prefix_of_array(x: [u32; 4]) -> [u32; 2] {
+    // CHECK-LABEL: fn prefix_of_array
+    // CHECK: _2 = copy _1 as {{.+}}::Transmute<[u32; 4], [u32; 2]> (Transmute);
+    // CHECK: _3 = move (_2.1: {{.+}}::ManuallyDrop<[u32; 2]>);
+    // CHECK: _0 = copy _3 as [u32; 2] (Transmute);
+    unsafe { transmute_prefix(x) }
+}
+
+#[repr(C, align(64))]
+struct Align64<T>(T);
+
+// EMIT_MIR transmutes.pad_for_alignment.PreCodegen.after.mir
+pub fn pad_for_alignment(x: u32) -> Align64<u32> {
+    // CHECK-LABEL: fn pad_for_alignment
+    // CHECK: _2 = copy _1 as {{.+}}::ManuallyDrop<u32> (Transmute);
+    // CHECK: _3 = {{.+}}::Transmute::<u32, Align64<u32>> { a: copy _2 };
+    // CHECK: _0 = move _3 as Align64<u32> (Transmute);
+    unsafe { transmute_prefix(x) }
+}
+
+// EMIT_MIR transmutes.forget_at_home.PreCodegen.after.mir
+pub fn forget_at_home(x: String) {
+    // CHECK-LABEL: fn forget_at_home
+    // CHECK: bb0:
+    // CHECK-NEXT: return;
+    unsafe { transmute_prefix(x) }
+}

